lms/resources/_js_config.py

```diff
@@ -77,6 +79,42 @@ def enable_content_item_selection_mode(self):
         """
         self.config["mode"] = "content-item-selection"
 
+    def maybe_set_focused_user(self):
+        """
+        Configure the Hypothesis client to focus on a particular user.
+
+        If there is a focused_user request param then add the necessary
+        Hypothesis client config to get the client to focus on the particular
+        user identified by the focused_user param, showing only that user's
+        annotations and not others.
+
+        In practice the focused_user param is only ever present in Canvas
+        SpeedGrader launches. We add a focused_user query param to the
+        SpeedGrader LTI launch URLs that we submit to Canvas for each student
+        when the student launches an assignment. Later, Canvas uses these URLs
+        to launch us when a teacher grades the assignment in SpeedGrader.
+
+        In theory, though, the focused_user param could work outside of Canvas
+        as well if we ever want it to.
+
+        """
+        focused_user = self._request.params.get("focused_user")
+
+        if not focused_user:
+            return
+
+        self._hypothesis_client["focus"] = {"user": {"username": focused_user}}
+
+        # Unfortunately we need to pass the user's current display name to the
+        # Hypothesis client, and we need to make a request to the h API to
+        # retrieve that display name.
+        try:
+            display_name = self._h_api.get_user(focused_user).display_name
+        except HAPIError:
+            display_name = "(Couldn't fetch student name)"
+
+        self._hypothesis_client["focus"]["user"]["displayName"] = display_name
```

Comment:
I had a go at refactoring it in this way, but I think it just ends up spreading the `focused_user` stuff in three separate places instead of two:

1. In the `record_canvas_speedgrader_submission()` API view, where we add the `focused_user` query param to the launch URL that we submit to SpeedGrader. (And of course the `h_username` that becomes the `focused_user` gets passed to this API by the JavaScript, and it gets passed to the JavaScript in the `submissionParams` part of the JS config.)

2. In `BasicLTILaunchViews.__init__()`, where we read the `focused_user` from `request.params` and either do nothing if it's not there, or if `focused_user` is there then fetch the `display_name` from `h_api` and then call `JSConfig`

3. In `JSConfig` where we actually add the `focused_user` to the JavaScript config where it's needed, when `BasicLTILaunchViews.__init__()` calls us.

It's unclear to me whether it's nicer to split the logic among three places like that, or whether the current solution (which collapses 2 and 3 into one place in `JSConfig`) is nicer.

There's going to be quite a few things in `JSConfig` that read quite a few request params and make decisions based on them, so that I think request params might form a pretty core part of what our JavaScript config logic is, and I think refactoring all of those to move the request param-based logic into the views, with the views deciding whether or not to call `JSConfig` methods and passing the params into them, might just mean that the logic about how our `JSConfig` works is spread around instead of in one place. So far, at various different times and under different circumstances `JSConfig` is going to have to read and sometimes make decisions based on these params: `custom_canvas_course_id`, `context_id`, `context_title`, `resource_link_id`, `resource_link_title`, `focused_user`, `lis_result_sourcedid`, `lis_outcome_service_url`, `oauth_consumer_key`, `tool_consumer_info_product_family_code`.

It won't in the end get rid of `maybe_set_focused_user()` either. It's just that `maybe_set_focused_user()` has now been moved from `JSConfig` into `BasicLTILaunchViews` (and rather than "maybe set focused user" it's "maybe ask `JSConfig` to set focused user"). I would probably still have a `maybe_set_focused_user()` method, it's just be a private helper method of `BasicLTILaunchViews` rather than a method of `JSConfig`, because I wouldn't want to inline the ~10 lines of code into `BasicLTILaunchViews.__init__()`, I'd still want them to be a separate method. (Basically the "maybe" logic is still there either way and is preexisting, the `maybe_set_focused_user()` method name just gives it a name.)

Reply:
Yeah, it feels like a no win to me, so why not have it in fewer places. It does have that spooky action at a distance feeling about it because Pyramid is so global variable based. :man_shrugging: